### PR TITLE
Export API

### PR DESCRIPTION
**Pull request by wsugarman**
## Description
- [x] Add new `ExportController` that defines the export API
- [x] Add new data models for the identifier source and azure blob sink
- [x] Add new DICOM operation enum value for export

## Related issues
Resolves [AB#90230](https://microsofthealth.visualstudio.com/Health/_workitems/edit/90230)

## Testing
New unit tests

### CODE REVIEW

File: src/Microsoft.Health.Dicom.Core/Serialization/ConfigurationJsonConverter.cs [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+using System.Collections.Generic;
+using System.Globalization;
+using System.Linq;
+using System.Text.Json;
+using System.Text.Json.Serialization;
+using Microsoft.Extensions.Configuration;
+
+namespace Microsoft.Health.Dicom.Core.Serialization;
+
+internal sealed class ConfigurationJsonConverter : JsonConverter<IConfiguration>
```
Comment (pengchen0692) [inline src/Microsoft.Health.Dicom.Core/Serialization/ConfigurationJsonConverter.cs:16]: I think it's quite creative to utilize `IConfiguration` to represent any customer input. 
I'm not expert on `IConfiguration`, not sure if there is any potential problems. 
I'll give an approve on this PR as it's under feature flag.

**Review by pengchen0692 — APPROVED** [review]
I  think it's quite creative to utilize IConfiguration to represent any customer input.
I'm not expert on IConfiguration, not sure if there is any potential problems.
I'll give an approve on this PR as it's under feature flag.

File: src/Microsoft.Health.Dicom.Api/Controllers/ExportController.cs [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+using Microsoft.Health.Dicom.Core.Models.Export;
+using Microsoft.Health.Dicom.Core.Web;
+using Microsoft.Health.Operations;
+
+namespace Microsoft.Health.Dicom.Api.Controllers;
+
+/// <summary>
+/// Represents an API controller for export operations.
+/// </summary>
+[ApiVersion("1.0-prerelease")]
```
Comment (smithago) [inline src/Microsoft.Health.Dicom.Api/Controllers/ExportController.cs:34]: Do we need to support this on prerelease version?

File: src/Microsoft.Health.Dicom.Api/Controllers/ExportController.cs [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+    /// <see cref="OperationReference"/> detailing the new export operation instance. Otherwise, the status code
+    /// provides details as to why the request failed.
+    /// </returns>
+    [HttpPost]
+    [BodyModelStateValidator]
+    [Produces(KnownContentTypes.ApplicationJson)]
+    [Consumes(KnownContentTypes.ApplicationJson)]
+    [ProducesResponseType(typeof(OperationReference), (int)HttpStatusCode.Accepted)]
+    [ProducesResponseType((int)HttpStatusCode.BadRequest)]
+    [Route(KnownRoutes.ExportInstancesRoute)]
```
Comment (smithago) [inline src/Microsoft.Health.Dicom.Api/Controllers/ExportController.cs:79]: can we only support v1 versioned route?

File: src/Microsoft.Health.Dicom.Api/Controllers/ExportController.cs (hunk 2 above)
Comment (wsugarman) [reply src/Microsoft.Health.Dicom.Api/Controllers/ExportController.cs:34]: Ah you're right! We can probably leave it

File: src/Microsoft.Health.Dicom.Api/Controllers/ExportController.cs [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+    [VersionedRoute(KnownRoutes.ExportInstancesRoute)]
+    [PartitionRoute(KnownRoutes.ExportInstancesRoute)]
+    [VersionedPartitionRoute(KnownRoutes.ExportInstancesRoute)]
+    [AuditEventType(AuditEventSubType.Export)]
+    public async Task<IActionResult> ExportInstancesAsync([Required][FromBody] ExportSpecification spec)
+    {
+        EnsureArg.IsNotNull(spec, nameof(spec));
```
Comment (smithago) [inline src/Microsoft.Health.Dicom.Api/Controllers/ExportController.cs:86]: Does the model validator already check this?

File: src/Microsoft.Health.Dicom.Core/Models/DicomIdentifier.cs [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+                if (parts.Length == 3)
+                {
+                    sopInstanceUid = parts[2];
+                    UidValidation.Validate(sopInstanceUid, nameof(SopInstanceUid));
+                }
+            }
+        }
+        catch (InvalidIdentifierException iie)
+        {
+            throw new FormatException(string.Format(CultureInfo.CurrentCulture, DicomCoreResource.InvalidDicomIdentifier, value), iie);
```
Comment (smithago) [inline src/Microsoft.Health.Dicom.Core/Models/DicomIdentifier.cs:223]: Is this caught has know exception in the Exceptionmiddleware? OR does this need to be wrapped on a validationException?

File: src/Microsoft.Health.Dicom.Core/Serialization/Newtonsoft/ConfigurationJsonConverter.cs [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+// -------------------------------------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Globalization;
+using System.Linq;
+using EnsureThat;
+using Microsoft.Extensions.Configuration;
+using Newtonsoft.Json;
```
Comment (smithago) [inline src/Microsoft.Health.Dicom.Core/Serialization/Newtonsoft/ConfigurationJsonConverter.cs:13]: Can't we use System.Text.Json?

File: src/Microsoft.Health.Dicom.Api/Controllers/ExportController.cs (hunk 4 above)
Comment (wsugarman) [reply src/Microsoft.Health.Dicom.Api/Controllers/ExportController.cs:86]: It does, but that would only occur if you invoke the method through the web, and the analyzer doesn't detect that. It worries users may invoke the method directly lol

File: src/Microsoft.Health.Dicom.Core/Models/DicomIdentifier.cs (hunk 5 above)
Comment (wsugarman) [reply src/Microsoft.Health.Dicom.Core/Models/DicomIdentifier.cs:223]: Oh, good catch. I'll check this

File: src/Microsoft.Health.Dicom.Core/Serialization/Newtonsoft/ConfigurationJsonConverter.cs (hunk 6 above)
Comment (wsugarman) [reply src/Microsoft.Health.Dicom.Core/Serialization/Newtonsoft/ConfigurationJsonConverter.cs:13]: The Durable Task framework takes a dependency on Newtonsoft.Json for its serialization, but one day I'll go make that change too 😛

**Review by pengchen0692 — APPROVED** [review]

**Review by pengchen0692 — APPROVED** [review]

**Pull request merged**